sailfishusermanager.cpp

```diff
@@ -0,0 +1,429 @@
+            line = droidDefault.readLine();
+            if (!line.isEmpty()) {
+                if (!addUserToGroup(user, line.trimmed())) {
+                    droidDefault.close();
+                    return false;
+                }
+            }
+        }
+        droidDefault.close();
+    }
+    QFile droidInet("/usr/lib/droid/hw-group.d/inet.list");
+    if (droidInet.open(QIODevice::ReadOnly)) {
+        while (!droidInet.atEnd()) {
+            line = droidInet.readLine();
+            if (!line.isEmpty()) {
+                if (!addUserToGroup(user, line.trimmed())) {
+                    droidInet.close();
+                    return false;
+                }
+            }
+        }
+        droidInet.close();
+    }
+
+    return true;
+}
+
+bool SailfishUserManager::makeHome(const QString &user)
+{
+    QString destination = QString(USER_HOME).arg(user);
+    QDir source(SKEL_DIR);
+    if (!source.mkdir(destination)) {
+        qWarning() << "User home already exists";
+        return false;
+    }
+
+    struct passwd *pw = getpwnam(user.toLocal8Bit());
+    if (!pw) {
+        qWarning() << "User not found";
+        return false;
+    }
+
+    if (chown(destination.toLocal8Bit(), pw->pw_uid, pw->pw_gid)) {
+        qWarning() << "Home directory ownership change failed";
+        return false;
+    }
+    if (chmod(destination.toLocal8Bit(), HOME_MODE)) {
+        qWarning() << "Home directory permissions change failed";
+        return false;
+    }
+
+    foreach (const QString& file, source.entryList(QDir::Files | QDir::NoDotAndDotDot | QDir::Hidden)) {
+        QString destFile = QString("%1/%2").arg(destination).arg(file);
+        if (!QFile::copy(QString("%1/%2").arg(source.path()).arg(file), destFile)) {
+            qWarning() << "Failed to copy file";
+            return false;
+        }
+        if (chown(destFile.toLocal8Bit(), pw->pw_uid, pw->pw_gid)) {
+            qWarning() << "Failed to change file ownership";
+            return false;
+        }
+    }
+
+    return true;
+}
+
+void SailfishUserManager::addUser(const QString& user, const QString& name, const QDBusMessage& message)
+{
+    m_timer->start();
+
+    if (QFile::exists(QString(USER_HOME).arg(user))) {
+        qWarning() << "Home directory already exists";
+        QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, "Home directory already exists"));
+        return;
+    }
+
+    if (!addGroup(user)) {
+        QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, "Creating user group failed"));
+        return;
+    }
+
+    struct lu_error *error = nullptr;
+    struct lu_context *context = lu_start(NULL, lu_user, NULL, NULL, lu_prompt_console, NULL, &error);
+    if (!context) {
+        qWarning() << "Error creating context:" << lu_strerror(error);
+        QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, lu_strerror(error)));
```

Comment:
I think inheriting `QDBusContext` and using `sendErrorReply()` would look nicer, no need for the `message` argument (if the message is needed for other reasons, like authorization, `message()` can be used). Also the introspection data would be compatible with other systems as well.

Reply:
Yeah, this message thing looks like some obscure QDBus specific hackery. I would vote for taking it out, too.

Reply:
I would also vote for not using Qt in daemons unless it's really necessary, but that's obviously too much to ask :D

Reply:
That would look nicer. This should also define its own errors instead of always returning QDBusError::Failed (org.freedesktop.DBus.Error.Failed) which tells nothing about what went wrong. There are also some possibly more appropriate errors to use in QDBusError like InvalidArgs or maybe InternalError if administrator can't do anything about it.

Reply:
I'm not sure, but it may even make sense to throw protocol specific errors, to let the client distinguish those from plain D-Bus errors. That may be an overkill, though - why would a client be interested in what exactly went wrong on the service side?

Reply:
I added those QDBusMessage arguments to slots so that I could answer with an error, but I'll try that QDBusContext if I can get that to work. Otherwise need to ditch automatic qdbusxml2cpp.